sqlstore/src/main/java/uk/ac/bbsrc/tgac/miso/db/migration/V8520__Attachments.java

```diff
@@ -0,0 +1,177 @@
+
+  private enum EntityType {
+    PROJECT("project", "Project", "projectId", "Project_Attachment"),
+    SERVICE_RECORD("servicerecord", "ServiceRecord", "recordId", "ServiceRecord_Attachment");
+    
+    private final String dirName;
+    private final String table;
+    private final String primaryKey;
+    private final String joinTable;
+    
+    private EntityType(String dirName, String table, String primaryKey, String joinTable) {
+      this.dirName = dirName;
+      this.table = table;
+      this.primaryKey = primaryKey;
+      this.joinTable = joinTable;
+    }
+
+    public String getDirName() {
+      return dirName;
+    }
+
+    public String getTable() {
+      return table;
+    }
+
+    public String getPrimaryKey() {
+      return primaryKey;
+    }
+
+    public String getJoinTable() {
+      return joinTable;
+    }
+  }
+
+  private Connection connection;
+  private String filesDir;
+  private Path basePath;
+
+  @Override
+  public void migrate(Connection connection) throws Exception {
+    log.info("Scanning existing attachments to save in database...");
+
+    this.connection = connection;
+    filesDir = getFilesDir();
+    if (LimsUtils.isStringEmptyOrNull(filesDir)) {
+      throw new IllegalStateException("MISO_FILES_DIR not set");
+    }
+    if (!filesDir.endsWith(File.separator)) {
+      filesDir += File.separator;
+    }
+    basePath = Paths.get(filesDir);
+
+    log.info("scanning directory {}", filesDir);
+    try (PreparedStatement insertStatement = connection.prepareStatement(
+        "INSERT INTO Attachment(filename, path, creator, created)"
+            + " VALUES (?, ?, (SELECT userId FROM User WHERE loginName = 'admin'), NOW());",
+        Statement.RETURN_GENERATED_KEYS)) {
+      for (EntityType type : EntityType.values()) {
+        scanTypeDir(type, insertStatement);
+      }
+    }
+
+    dropTempTable();
+  }
+
+  private void scanTypeDir(EntityType type, PreparedStatement insertStatement) throws SQLException {
+    File dir = new File(filesDir + type.getDirName());
+    if (dir.exists()) {
+      log.info("scanning directory {}", dir.getAbsolutePath());
+      try (PreparedStatement getStatement = connection
+          .prepareStatement(String.format("SELECT * FROM %s WHERE %s = ?", type.getTable(), type.getPrimaryKey()));
+          PreparedStatement joinStatement = connection.prepareStatement(
+              String.format("INSERT INTO %s(%s, attachmentId) VALUES (?, ?);", type.getJoinTable(), type.getPrimaryKey()))) {
+        for (File typeFile : dir.listFiles()) {
+          processObjectDir(type, typeFile, getStatement, insertStatement, joinStatement);
+        }
+      }
+    }
+  }
+
+  private void processObjectDir(EntityType type, File dir, PreparedStatement getStatement, PreparedStatement insertStatement,
+      PreparedStatement joinStatement) throws SQLException {
+    log.info("scanning directory " + dir.getAbsolutePath());
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Format specifiers should be used instead of string concatenation. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)